graph/pull_v2.go

```diff
@@ -0,0 +1,369 @@
+	"github.com/docker/docker/trust"
+	"github.com/docker/docker/utils"
+	"github.com/docker/libtrust"
+)
+
+type v2Puller struct {
+	*TagStore
+	endpoint registry.APIEndpoint
+}
+
+func (p *v2Puller) Pull(canonicalImageName, name, tag string, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (fallback bool, err error) {
+	// TODO(tiborvass): was ReceiveTimeout
+	repo, err := NewV2Repository(name, p.endpoint, imagePullConfig.MetaHeaders, imagePullConfig.AuthConfig)
+	if err != nil {
+		logrus.Debugf("Error getting v2 registry: %v", err)
+		return true, err
+	}
+
+	if err := p.pullV2Repository(repo, canonicalImageName, tag, imagePullConfig, sf); err != nil {
+		if registry.ContinueOnError(err) {
+			logrus.Debugf("Error trying v2 registry: %v", err)
+			return true, err
+		}
+		return false, err
+	}
+	return false, nil
+}
+
+func (p *v2Puller) pullV2Repository(repo distribution.Repository, image, tag string, imagePullConfig *ImagePullConfig, sf *streamformatter.StreamFormatter) (err error) {
+	var tags []string
+	taggedName := image
+	if len(tag) > 0 {
+		tags = []string{tag}
+		taggedName = utils.ImageReference(image, tag)
+	} else {
+		var err error
+		tags, err = repo.Manifests().Tags()
+		if err != nil {
+			return err
+		}
+
+	}
+
+	c, err := p.poolAdd("pull", taggedName)
+	if err != nil {
+		if c != nil {
+			// Another pull of the same repository is already taking place; just wait for it to finish
+			sf.FormatStatus("", "Repository %s already being pulled by another client. Waiting.", image)
+			<-c
+			return nil
+		}
+		return err
+	}
+	defer p.poolRemove("pull", taggedName)
+
+	var layersDownloaded bool
+	for _, tag := range tags {
+		// pulledNew is true if either new layers were downloaded OR if existing images were newly tagged
+		// TODO(tiborvass): should we change the name of `layersDownload`? What about message in WriteStatus?
+		pulledNew, err := p.pullV2Tag(repo, image, tag, taggedName, imagePullConfig.OutStream, sf)
+		if err != nil {
+			return err
+		}
+		layersDownloaded = layersDownloaded || pulledNew
+	}
+
+	WriteStatus(taggedName, imagePullConfig.OutStream, sf, layersDownloaded)
+
+	return nil
+}
+
+func (p *v2Puller) pullV2Tag(repo distribution.Repository, localName, tag, taggedName string, out io.Writer, sf *streamformatter.StreamFormatter) (bool, error) {
+	logrus.Debugf("Pulling tag from V2 registry: %q", tag)
+
+	// downloadInfo is used to pass information from download to extractor
+	type downloadInfo struct {
+		img     *Image
+		tmpFile *os.File
+		digest  digest.Digest
+		layer   distribution.ReadSeekCloser
+		size    int64
+		err     chan error
+	}
+
+	manifest, err := repo.Manifests().GetByTag(tag)
+	if err != nil {
+		return false, err
+	}
+	verified, err := p.validateManifest(manifest, tag)
+	if err != nil {
+		return false, fmt.Errorf("error verifying manifest keys: %v", err)
+	}
+	if verified {
+		// TODO(tiborvass): isn't this misleading since the layers have not been verified yet?
```

Comment:
This is just intended to inform on whether the manifest is verified. If the manifest is verified then every other value downloaded will also be verified. If any verification fails, the operation will fail. Yes the operation can fail despite the manifest being verified, this could still be used for informative purposes about the manifest which provided the content addresses which failed.